### ISSUE TITLE
No backpressure when converting to publisher and back

### ISSUE THREAD

**Issue by jchapuis** [monix/monix#352]
I'm setting up backpressure-enabled streams over grpc and I would like to expose the streams in the public api using standard reactive streams types (i.e. `Publisher[T]`). 

In order to do so I'm converting my stream `Observable` using `toReactivePublisher()`, and then in my client/test environment consuming that once again with monix by converting it back to an `Observable` using `Observable.fromReactivePublisher()`. 

However, I'm observing a different behavior in terms of backpressure compared to the fully native monix version: going through this reactive streams "bridge" seems to make it eager, and it consumes all the stream even though I'm using delay operators down the line. Any idea what could cause it to ignore backpressure and this difference in behavior? Thanks

**Comment by alexandru** [comment]
@jchapuis converting into a reactive publisher or from a reactive publisher involves buffering.

So what happens is that a fixed size buffer gets created and the source is free to push events for as long as that buffer has room in it. And then back-pressuring happens after the buffer gets full.

**Comment by jchapuis** [comment]
@alexandru thanks! any specific reason for the buffering? is there any way to configure the size of that buffer?

**Comment by alexandru** [comment]
Currently the [ReactiveObservable](https://github.com/monix/monix/blob/v2.2.4/monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/ReactiveObservable.scala) (the implementation that converts `org.reactivestreams.Publisher` into `Observable` is doing `subscriber.toReactive`, which initiates a [SubscriberAsReactiveSubscriber](https://github.com/monix/monix/blob/v2.2.4/monix-reactive/shared/src/main/scala/monix/reactive/internal/rstreams/SubscriberAsReactiveSubscriber.scala) using a buffer size set by the `Scheduler`'s execution model, see: https://monix.io/docs/2x/execution/scheduler.html#execution-model

For now, if you really, really want this, you can configure the scheduler passed when building that publisher. I don't recommend that though. 

You could also do your own helper:

```scala
def convertToObservable[A](publisher: Publisher[A], bufferSize: Int): Publisher[A] = 
  new Observable[A] {
    def unsafeSubscribeFn(subscriber: Subscriber[A]): Cancelable = {
      val sub = subscriber.toReactive(bufferSize)
      val subscription = SingleAssignmentSubscription()

      publisher.subscribe(new reactivestreams.Subscriber[A] {
        def onNext(t: A): Unit = sub.onNext(t)
        def onComplete(): Unit = sub.onComplete()
        def onError(t: Throwable): Unit = sub.onError(t)

        def onSubscribe(s: Subscription): Unit = {
          subscription := s
          sub.onSubscribe(s)
        }
      })

      subscription
    }
  }
```

So it's all in that created `Subscriber`. We have a buffer size parameter there, but we don't have it on `Observable.fromReactivePublisher`.

A PR for an overload accepting a buffer size is is welcome.

**Comment by jchapuis** [comment]
Ok thanks! I actually did try exactly what you are proposing earlier, with my own helper because I had spotted that `toReactive` overload that accepts a buffer size. However it didn't seem to change the behavior. I could try again if it works do a PR indeed. 

Btw. turning the publisher created with Monix into a RxJava2 Flowable works like a charm in Java, with backpressure.

**Comment by jchapuis** [comment]
Ok it doesn't seem to work. However I've spotted a possible reason for this in `SubscriberAsReactiveSubscriber`:

```scala
private[this] val buffer: RSubscriber[T] =
    SyncSubscriberAsReactiveSubscriber(
      BufferedSubscriber.synchronous(subscriber, Unbounded),
      requestCount = requestCount
    )
```
Notice the `Unbounded` parameter here, would that be it?

**Comment by alexandru** [comment]
@jchapuis sorry for the silence, I've been busy and couldn't investigate if this works as designed or not.

I'm planning to cut a new 2.3.0 release ASAP, sometime this week, however I promise to have this issue sorted out until then.

**Comment by jchapuis** [comment]
@alexandru, thanks for your reply, no problem, good to know!



On 2 May 2017, at 21:15, Alexandru Nedelcu <notifications@github.com<mailto:notifications@github.com>> wrote:


@jchapuis<https://github.com/jchapuis> sorry for the silence, I've been busy and couldn't investigate if this works as designed or not.

I'm planning to cut a new 2.3.0 release ASAP, sometime this week, however I promise to have this issue sorted out until then.

**Comment by alexandru** [comment]
@jchapuis there's indeed a problem with that unbounded buffer. Well, it's not the unbounded-ness that's the issue, but how we keep requesting for stuff.

I'm working on a new implementation. Stay tuned.

**Comment by alexandru** [comment]
Fix is on master.

**Comment by shengc** [comment]
Is this issue actually fixed ? It seems like it is still there for `3.0.0-RC1`. I set `requestCount = 1`, yet yourkit showed the memory taken by the process kept growing until it died with OOM.
